compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

```diff
@@ -781,11 +781,32 @@ class ClassfileParser(
 
       if (scan(tpnme.TASTYATTR)) {
         val attrLen = in.nextInt
-        if (attrLen == 0) {
-          // A tasty attribute implies the existence of the .tasty file
-          val file = new PlainFile(io.File(classfile.jpath).changeExtension("tasty"))
-          if (file.exists) return unpickleTASTY(new AbstractFileReader(file).nextBytes(file.sizeOption.get))
-          else ctx.error("Could not find " + file)
+        if (attrLen == 0) { // A tasty attribute implies the existence of the .tasty file
+          def readBytes(file: io.Path): Array[Byte] = {
+            val plainFile = new PlainFile(file)
+            if (plainFile.exists) new AbstractFileReader(plainFile).nextBytes(plainFile.sizeOption.get)
+            else {
+              ctx.error("Could not find " + file)
+              Array.empty
+            }
+          }
+          val tastyBytes = classfile match {
```

Comment:
I propose something along the line of:
```scala
val tastyBytes = classfile.underlyingSource match {
  case None =>
    readBytes(File(classfile.jpath).changeExtension("tasty"))
  case Some(jar) => // We are in a jar
    val jarArchive = JarArchive.open(File(jar.jpath))
    val tastyFile = File(jarFile.jpath).resolve(classfile.name).changeExtension("tasty")
    val bytes = readBytes(tastyFile)
    jarArchive.close()
    bytes
}
```
It's weird to pattern match on `FileZipArchive#LeakyEntry`. This is internal implementation detail, should probably be private.

Reply:
Also can you add a comment: `// TODO: simplify when #3552 is fixed`

Reply:
I have updated the code snippet: `underlyingSource` seems to return `None` only for virtual files. In the case there is no `underlyingSource`, we cannot do anything...